### PR TITLE
Preliminary support for `_specialize(exported: true, ...)`

### PR DESCRIPTION
**Pull request by aschwaighofer**
Adds support for `_specialize(exported: true, ...)`

This attribute allows to define a pre-specialized entry point of a
generic function in a library.

The following definition provides a pre-specialized entry point for
`genericFunc(_:)` for the parameter type `Int` that clients of the
library can call.

```
@_specialize(exported: true, where T == Int)
public func genericFunc<T>(_ t: T) { ... }
```

Pre-specializations of internal `@inlinable` functions are allowed.

```
@usableFromInline
internal struct GenericThing<T> {
  @_specialize(exported: true, where T == Int)
  @inlinable
  internal func genericMethod(_ t: T) {
  }
}
```

There is syntax to pre-specialize a method from a different module.

```
import ModuleDefiningGenericFunc

@_specialize(exported: true, target: genericFunc(_:), where T == Double)
func prespecialize_genericFunc(_ t: T) { fatalError("dont call") }

```

Specially marked extensions allow for pre-specialization of internal
methods accross module boundries (respecting `@inlinable` and
`@usableFromInline`).

```
import ModuleDefiningGenericThing
public struct Something {}

@_specializeExtension
extension GenericThing {
  @_specialize(exported: true, target: genericMethod(_:), where T == Something)
  func prespecialize_genericMethod(_ t: T) { fatalError("dont call") }
}
```

Adds support for marking a _specialize attribute as SPI 

```
  @_specialize(exported: true, spi: SPIGroupName, where T == Int)
  public func myFunc() { } 
```

The specialized entry point is only visible for modules that import
using `_spi(SPIGroupName) import ModuleDefiningMyFunc `.

rdar://64993425

### CODE REVIEW

File: lib/SIL/IR/SILDeclRef.cpp [hunk 1]
```diff
@@ -221,6 +228,12 @@ bool SILDeclRef::isImplicit() const {
 }
 
 SILLinkage SILDeclRef::getLinkage(ForDefinition_t forDefinition) const {
+
+  // Prespecializations are public.
```
Comment (slavapestov) [inline lib/SIL/IR/SILDeclRef.cpp:232]: Only specializations of public symbols though, right?

File: include/swift/SIL/SILDeclRef.h [hunk 2]
```diff
@@ -151,6 +153,8 @@ struct SILDeclRef {
   /// The derivative function identifier.
   AutoDiffDerivativeFunctionIdentifier *derivativeFunctionIdentifier = nullptr;
 
+  GenericSignature specializedSignature;
```
Comment (slavapestov) [inline include/swift/SIL/SILDeclRef.h:156]: Can we stick this in a union with derivativeFunctionIdentifier to save space?

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -631,9 +631,6 @@ bool Parser::parseSpecializeAttributeArguments(
       if (ParamLabel == "exported") {
         auto trueLoc = Tok.getLoc();
         bool isTrue = consumeIf(tok::kw_true);
-        if (isTrue) {
-          diagnose(trueLoc, diag::attr_specialize_export_true_no_op);
```
Comment (slavapestov) [inline lib/Parse/ParseDecl.cpp:635]: I think you can remove this diagnostic now

File: lib/SIL/Utils/GenericSpecializationMangler.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      CalleeInterfaceToSpecializedInterfaceMapping;
+
+  // TODO: Can this be simplified?
+  //
+
+  // Callee's generic signature and specialized generic signature
+  // use the same set of generic parameters, i.e. each generic
+  // parameter should be mapped to itself.
+  for (auto GP : CalleeGenericSig->getGenericParams()) {
+    CalleeInterfaceToSpecializedInterfaceMapping[GP] = Type(GP);
```
Comment (slavapestov) [inline lib/SIL/Utils/GenericSpecializationMangler.cpp:120]: Do you really need CalleeInterfaceToSpecializedInterfaceMapping at all then? The lookup() below can be replaced with just 'type' itself, no?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 5]
```diff
@@ -1762,9 +1762,12 @@ void ReabstractionInfo::finishPartialSpecializationPreparation(
...
                                      bool isWholeModule, SILFunction *Callee,
-                                     GenericSignature SpecializedSig)
-    : TargetModule(targetModule), isWholeModule(isWholeModule) {
-  Serialized = Callee->isSerialized();
+                                     GenericSignature SpecializedSig,
+                                     bool isPrespecialization)
+    : TargetModule(targetModule), isWholeModule(isWholeModule),
+      isPrespecialization(isPrespecialization) {
+  Serialized =
+      this->isPrespecialization ? IsNotSerialized : Callee->isSerialized();
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp:1770]: Why can't we serialize the pre-specialization if the original function is serialized?

**Review by slavapestov — COMMENTED** [review]
Cool! Can this eventually be used to generate SwiftOnoneSupport too?

File: include/swift/SIL/SILDeclRef.h (hunk 2 above)
Comment (aschwaighofer) [reply include/swift/SIL/SILDeclRef.h:156]: Good point.

File: lib/Parse/ParseDecl.cpp (hunk 3 above)
Comment (aschwaighofer) [reply lib/Parse/ParseDecl.cpp:635]: I did.

File: lib/SIL/Utils/GenericSpecializationMangler.cpp (hunk 4 above)
Comment (aschwaighofer) [reply lib/SIL/Utils/GenericSpecializationMangler.cpp:120]: Yes. Likely. I have copied this code from the FunctionSignaturePartialSpecializer and have not gone back to simplify it.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 5 above)
Comment (aschwaighofer) [reply lib/SILOptimizer/Utils/Generics.cpp:1770]: This is intentional. Pre-specialization as implemented in this patch get their own symbol/mangling to make them easily identifiable for tooling. The intent of exported pre-specializations is to reduce code size. When using the pre-specializated symbol I don't want the code to be available in the client module for further optimization.

File: lib/SIL/IR/SILDeclRef.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/SIL/IR/SILDeclRef.cpp:232]: Pre-specialization are intentionally always public.

For example, there is going to be a public entry point for the specialized version of:

```
@_alwaysEmitIntoClient
@_specialize(exported: true, where T == Int)
internal func testGeneric<T>(t: T) {
  print(t)
}
```

**Comment by aschwaighofer** [comment]
> Can this eventually be used to generate SwiftOnoneSupport too?

I think so.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
Please test with following PR:
https://github.com/apple/swift-syntax/pull/241

@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test source compatibility

**Comment by aschwaighofer** [comment]
The compiler failure in the source compatibility run also occurs without this change in: https://ci.swift.org/job/swift-main-source-compat-suite/5513/artifact/swift-source-compat-suite/FAIL_swift-nio-ssl_5.0_BuildSwiftPackage.log

**Pull request merged**